extensions/notebook/src/book/bookModel.ts

```diff
@@ -0,0 +1,221 @@
+import { BookTreeItem, BookTreeItemType } from './bookTreeItem';
+import { maxBookSearchDepth, notebookConfigKey } from '../common/constants';
+import * as path from 'path';
+import * as fs from 'fs-extra';
+import * as nls from 'vscode-nls';
+import { IJupyterBookToc, IJupyterBookSection } from '../contracts/content';
+
+const localize = nls.loadMessageBundle();
+//const exists = promisify(fs.exists);
+
+export class BookModel implements azdata.nb.NavigationProvider {
+	public BookPath: string;
+	private _bookItems: BookTreeItem[] = [];
+	private _allNotebooks = new Map<string, BookTreeItem>();
+	public openAsUntitled: boolean = false;
+	private _extensionContext: vscode.ExtensionContext;
+	private _tableOfContentPaths: string[] = [];
+	readonly providerId: string = 'BookNavigator';
+	// For testing
+	private _errorMessage: string;
+
+	constructor(bookPath: string, openAsUntitled: boolean, extensionContext: vscode.ExtensionContext) {
+		this.BookPath = bookPath;
+		this.openAsUntitled = openAsUntitled;
+		this._extensionContext = extensionContext;
+		extensionContext.subscriptions.push(azdata.nb.registerNavigationProvider(this));
+		//this.getTableOfContentFiles(this.BookPath).then(() => { this._books = this.readBooks();}, (err) => { console.log(err); });
+
+	}
+
+	public async initializeContents(): Promise<void> {
+		await this.getTableOfContentFiles(this.BookPath);
+		this.readBooks();
+		return;
+	}
+
+	public getAllBooks(): Map<string, BookTreeItem> {
+		return this._allNotebooks;
+	}
+
+	async getTableOfContentFiles(workspacePath: string): Promise<void> {
+		try {
+			let notebookConfig = vscode.workspace.getConfiguration(notebookConfigKey);
+			let maxDepth = notebookConfig[maxBookSearchDepth];
+			// Use default value if user enters an invalid value
+			if (maxDepth === undefined || maxDepth < 0) {
+				maxDepth = 5;
+			} else if (maxDepth === 0) { // No limit of search depth if user enters 0
+				maxDepth = undefined;
+			}
+
+			let p = path.join(workspacePath, '**', '_data', 'toc.yml').replace(/\\/g, '/');
+			let tableOfContentPaths = await glob(p, { deep: maxDepth });
+			this._tableOfContentPaths = this._tableOfContentPaths.concat(tableOfContentPaths);
+			let bookOpened: boolean = this._tableOfContentPaths.length > 0;
+			vscode.commands.executeCommand('setContext', 'bookOpened', bookOpened);
+		} catch (error) {
+			console.log(error);
+		}
+	}
+
+	public readBooks(): BookTreeItem[] {
+		for (const contentPath of this._tableOfContentPaths) {
+			let root = path.dirname(path.dirname(contentPath));
+			try {
+				const config = yaml.safeLoad(fs.readFileSync(path.join(root, '_config.yml'), 'utf-8'));
+				const tableOfContents = yaml.safeLoad(fs.readFileSync(contentPath, 'utf-8'));
+				let book = new BookTreeItem({
+					title: config.title,
+					root: root,
+					tableOfContents: { sections: this.parseJupyterSection(tableOfContents) },
+					page: tableOfContents,
+					type: BookTreeItemType.Book,
```

Comment:
if we're creating a new BookTreeItem, why do we need to pass in BookTreeItemType.Book?

Reply:
BookTreeItem can be either of the following : Book, Notebook, Markdown, ExternalLink.
In the readBooks() we initialize the parent node of the book so we have it as Book. And in the getSections() we get each file inside the Book where we specify the type appropriately.